File: 744-find-smallest-letter-greater-than-target/744-find-smallest-letter-greater-than-target.cpp
class Solution {
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        const int len = letters.size() - 1;
        int low = 0, high = len, mid;
        
        while (low <= high) {
            mid = ceil((high + low) / 2.0);
            
            if (letters[mid] <= target) 
                low = mid + 1;
            else if (letters[mid] > target) 
                high = mid - 1;
        }
        
        const int idx = max(low, high) > len ? 0 : max(low, high);
        return letters[idx];
    }
};

File: 744-find-smallest-letter-greater-than-target/744-find-smallest-letter-greater-than-target.go
func nextGreatestLetter(letters []byte, target byte) byte {
    low, high := 0, len(letters) - 1
    
    for low <= high {
        mid := (high + low) / 2
        
        if target < letters[mid] {
            high = mid - 1
        } else {
            low = mid + 1
        }
    }

    return letters[low % len(letters)]
}